pkg/containerd/cmd/service/prepare.go

```diff
@@ -4,18 +4,79 @@ package main
 // Update it in both places if you make changes
 
 import (
+	"fmt"
+	"os"
 	"path/filepath"
 	"syscall"
 )
 
 func prepare(path string) error {
+	// see if we are dealing with a read only or read write container
+	if _, err := os.Stat(filepath.Join(path, "lower")); err != nil {
+		return prepareRO(path)
+	}
+	return prepareRW(path)
+}
+
+func prepareRO(path string) error {
+	// make rootfs a mount point, as runc doesn't like it much otherwise
 	rootfs := filepath.Join(path, "rootfs")
 	if err := syscall.Mount(rootfs, rootfs, "", syscall.MS_BIND, ""); err != nil {
 		return err
 	}
-	// remount rw
-	if err := syscall.Mount("", rootfs, "", syscall.MS_REMOUNT, ""); err != nil {
+	return nil
+}
+
+func prepareRW(path string) error {
+	// mount a tmpfs on tmp for upper and workdirs
+	// make it private as nothing else should be using this
+	tmp := filepath.Join(path, "tmp")
+	if err := syscall.Mount("tmpfs", tmp, "tmpfs", 0, "size=10%"); err != nil {
```

Comment:
The hardcoding of 10% is a bit of a shame.

What about this: have moby tool write a list of mounts (as a JSON array) into some file (lets say `rootfs.json`) and this tool reads it and passes it (perhaps after API conversion) to `WithRootFS` when calling `container.NewTask`? Then the tool has some control over all this and we can more easily expose via the yml files etc?